IdentityCore/src/JWEResponse/MSIDJweResponse+EcdhAesGcm.m

```diff
@@ -0,0 +1,251 @@
+MSIDJWECryptoKeyResponseEncryptionAlgorithm const MSID_RESPONSE_ENCRYPTION_ALGORITHM_A256GCM = @"A256GCM";
+
+@implementation MSIDJweResponse (EcdhAesGcm)
+
+- (nullable NSDictionary *)decryptJweResponseWithPrivateStk:(nonnull SecKeyRef)privateStkRef
+                                                  jweCrypto:(nonnull MSIDJWECrypto *)jweCrypto
+                                                      error:(NSError * _Nullable __autoreleasing * _Nullable)error
+{
+    // 1. Check for necessary request parameters
+    NSData *apv = [NSData msidDataFromBase64UrlEncodedString:jweCrypto.apv.APV];
+    
+    if (!apv.length)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unexpected input parameters, no apv present in request JWE Crypto", nil, nil, nil, nil, nil, NO);
+        }
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Unexpected input parameters, no apv present in request JWE Crypto");
+        return nil;
+    }
+    
+    // 2. Check for necessary response parameters, epk, enc, apu
+    NSDictionary *epk = self.jweHeader[@"epk"];
+    
+    if (![epk count]
+        || ![epk isKindOfClass:[NSDictionary class]])
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unexpected server response, no epk present in JWE header", nil, nil, nil, nil, nil, NO);
+        }
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Unexpected server response, no epk present in JWE header, epk %@", epk);
+        return nil;
+    }
+    
+    NSString *encHeader = self.jweHeader[@"enc"];
+    NSString *apuHeader = self.jweHeader[@"apu"];
+    
+    if ([NSString msidIsStringNilOrBlank:encHeader]
+        || [NSString msidIsStringNilOrBlank:apuHeader])
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unexpected server response, no enc or apu present JWE header", nil, nil, nil, nil, nil, NO);
+        }
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Unexpected server response, no enc or apu present JWE header, enc %@, apu %@", encHeader, apuHeader);
+        return nil;
+    }
+    
+    if (![self IsJweResponseAlgorithmSupported:error])
+    {
+        return nil;
+    }
+    
+    // 3. Create key from server response
+    SecKeyRef serverKeyRef = [NSData createECCKeyFromEccJsonWebKey:epk error:error];
+    
+    if (!serverKeyRef)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Could not get EC public key from server JWE response payload.");
+        return nil;
+    }
+    
+    // 4. Calculate shared key through a key exchange
+    // It performs a Diffie-Hellman key exchange (kSecKeyAlgorithmECDHKeyExchangeStandard) using the Device Encryption public key and the Ephemeral private key.
+    NSData *sharedSecret = [self calculateSharedSecretUsingEcdhWithOtherPartyPublicEcKey:serverKeyRef
```

Comment:
I think we should clear these values right after used (derived key -> reset sharedScecret, etc...) That is the way eSTS implementation and I think it's good practice.

Reply:
Sure, cleared out sharedSecret and derived key after they are used.